runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/work/budget/EvenGetWorkBudgetDistributor.java

```diff
@@ -0,0 +1,99 @@
+
+import java.math.RoundingMode;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import org.apache.beam.runners.dataflow.worker.windmill.client.grpc.WindmillStreamSender;
+import org.apache.beam.sdk.annotations.Internal;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.ImmutableCollection;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.ImmutableMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Evenly distributes the provided budget across the available {@link WindmillStreamSender}(s). */
+@Internal
+public final class EvenGetWorkBudgetDistributor implements GetWorkBudgetDistributor {
+  private static final Logger LOG = LoggerFactory.getLogger(EvenGetWorkBudgetDistributor.class);
+  private final Supplier<GetWorkBudget> activeWorkBudgetSupplier;
+
+  public EvenGetWorkBudgetDistributor(Supplier<GetWorkBudget> activeWorkBudgetSupplier) {
+    this.activeWorkBudgetSupplier = activeWorkBudgetSupplier;
+  }
+
+  private static boolean isBelowFiftyPercentOfTarget(
+      GetWorkBudget remaining, GetWorkBudget target) {
+    return remaining.items() < (target.items() * 0.5) || remaining.bytes() < (target.bytes() * 0.5);
+  }
+
+  @Override
+  public void distributeBudget(
+      ImmutableCollection<WindmillStreamSender> streams, GetWorkBudget getWorkBudget) {
+    if (streams.isEmpty()) {
+      LOG.debug("Cannot distribute budget to no streams.");
+      return;
+    }
+
+    if (getWorkBudget.equals(GetWorkBudget.noBudget())) {
+      LOG.debug("Cannot distribute 0 budget.");
+      return;
+    }
+
+    Map<WindmillStreamSender, GetWorkBudget> desiredBudgets =
+        computeDesiredBudgets(streams, getWorkBudget);
+
+    for (Entry<WindmillStreamSender, GetWorkBudget> streamAndDesiredBudget :
+        desiredBudgets.entrySet()) {
+      WindmillStreamSender stream = streamAndDesiredBudget.getKey();
+      GetWorkBudget desired = streamAndDesiredBudget.getValue();
+      GetWorkBudget remaining = stream.remainingGetWorkBudget();
+      if (isBelowFiftyPercentOfTarget(remaining, desired)) {
+        GetWorkBudget adjustment = desired.subtract(remaining);
+        LOG.info("Adjusting budget for stream={} by {}", stream, adjustment);
+        stream.adjustBudget(adjustment);
+      }
+    }
+  }
+
+  private ImmutableMap<WindmillStreamSender, GetWorkBudget> computeDesiredBudgets(
+      ImmutableCollection<WindmillStreamSender> streams, GetWorkBudget totalGetWorkBudget) {
+    GetWorkBudget activeWorkBudget = activeWorkBudgetSupplier.get();
+    LOG.info("Current active work budget: {}", activeWorkBudget);
+    GetWorkBudget budgetPerStream =
+        GetWorkBudget.builder()
+            .setItems(
+                divide(
+                    totalGetWorkBudget.items() - activeWorkBudget.items(),
+                    streams.size(),
+                    RoundingMode.CEILING))
```

Comment:
rounding up here will drift upwards over the lifetime of the stream if the budget is a closed loop otherwise.

Reply:
added a comment and TODO to fix the non-determinism